Add multi key and key id support.

nigoroll opened the pull request:
This PR consists of two patches:
* The first adds support for accepting multiple keys
* The second adds support for named keys (with key ids)

Please refer to the individual commit messages and the updated documentation.

A changelog will be added to the PR if/when otherwise accepted.

nigoroll commented:
The first commit is #32

nigoroll commented:
If I read [the coverage report](https://github.com/Styria-Digital/django-rest-framework-jwt/pull/33#issuecomment-586701686) correctly, the delta stems from https://github.com/Styria-Digital/django-rest-framework-jwt/pull/33/commits/2e124f60e10ea841be33060bbc560f71ace06092 which, as explained in the commit message, I this focuses on python 3.7 for a good reason.

fitodic commented on docs/index.md:
```diff
@@ -179,6 +179,30 @@ This package uses the JSON Web Token Python implementation, [PyJWT](https://gith
+
+For a list or dict, the first element/value is used for
+signing. *NOTE: In python < 3.7 it is undefined which value from a
+dict is used for signing.*
```
Perhaps this note can explain why this behavior is so?

fitodic commented on docs/index.md:
````diff
@@ -179,6 +179,30 @@ This package uses the JSON Web Token Python implementation, [PyJWT](https://gith
+  ```
+
+  If a JWT is to be verified which contains a key id (kid header), only
+  the associated key is tried. Otherwise, or
````
Seeing its a list, I think this _Otherwise, or_ is a bit redundant. What are your thoughts on the matter?

fitodic commented on src/rest_framework_jwt/utils.py:
```diff
@@ -95,8 +95,23 @@ def jwt_get_username_from_payload_handler(payload):
-    return force_text(jwt.encode(payload, key, api_settings.JWT_ALGORITHM))
+    headers=None
+
+    algo = api_settings.JWT_ALGORITHM
```
Can we use `algorithm` instead?

fitodic commented on src/rest_framework_jwt/utils.py:
```diff
@@ -95,8 +95,23 @@ def jwt_get_username_from_payload_handler(payload):
+    elif isinstance(key,list):
+        key = key[0]
+
+    return force_text(jwt.encode(payload, key, algo, headers=headers))
```
`force_text` will start raising deprecation warnings on Django 3+.

nigoroll replied:
That was pre-existing code, replaced with `.decode()`. correct?

fitodic commented on docs/index.md:
````diff
@@ -190,19 +214,56 @@ Default is `None`.
+  keys, e.g.:
+
+  ```python
+  { "kid1": key1, "kid2": key2, ... }
````
If these settings only contain `kid`s, why use a dictionary instead of an iterable (tuple, list)?

nigoroll replied:
For efficient lookup by key id

fitodic commented on src/rest_framework_jwt/utils.py:
```diff
@@ -95,8 +95,23 @@ def jwt_get_username_from_payload_handler(payload):
+        kid, key = next(iter(key.items()))
+        headers = {"kid": kid}
+    elif isinstance(key,list):
+        key = key[0]
```
I would suggest splitting signing keys from verification keys to make it as explicit as possible. I would also suggest using `signing_key`.

nigoroll replied:
see top level discussion for suggestions, I'd appreciate your feedback

fitodic commented on src/rest_framework_jwt/utils.py:
```diff
@@ -105,17 +120,51 @@ def jwt_decode_token(token):
-        ]
-    )
+
+    algos = api_settings.JWT_ALGORITHM
```
As mentioned earlier, I would prefer using more explicit variable names.

nigoroll replied:
ok, renamed

fitodic commented on src/rest_framework_jwt/utils.py:
```diff
@@ -105,17 +120,51 @@ def jwt_decode_token(token):
+        keys = api_settings.JWT_PUBLIC_KEY
+
+    # if we got named keys and a key id is present, we insist to have
+    # that one, otherwise we try all
```
Why is this necessary?

nigoroll replied:
see top level discussion

fitodic commented on src/rest_framework_jwt/utils.py:
```diff
@@ -105,17 +120,51 @@ def jwt_decode_token(token):
+                issuer=api_settings.JWT_ISSUER, algorithms=[alg]
+            )
+        except (jwt.exceptions.InvalidSignatureError) as e:
+                ex = e
```
Why not raise the exception here?

nigoroll replied:
To try multiple keys for the case of no key id

fitodic commented on tests/views/test_authentication.py:
```diff
@@ -5,12 +5,41 @@
+    from cryptography.hazmat.backends import default_backend
+    from cryptography.hazmat.primitives.asymmetric import rsa
+
+    # DO NOT copy key_size=512 for any real code!
```
Why?

nigoroll replied:
because it's probably to be considered insecure for any real world application, yet I wanted to keep the test case efficient.

fitodic commented on tests/views/test_authentication.py:
```diff
@@ -5,12 +5,41 @@
+    prsa1 = srsa1.public_key()
+    prsa2 = srsa2.public_key()
+
+except:
```
This is a known [anti-pattern in Python](https://realpython.com/the-most-diabolical-python-antipattern/). Which exceptions are expected to be raised here?

If these keys need to be reused, shouldn't they be placed in a [pytest fixture](https://docs.pytest.org/en/latest/fixture.html)?

nigoroll replied:
thank you for helping out a python dummy :)

fitodic commented on tests/views/test_authentication.py:
```diff
@@ -159,3 +188,115 @@ def test_auth_cookie_settings(
+        srsa1
+    except:
+        skip("no rsa available")
+        return
```
This is what pytest fixtures are used for, to execute the setup.

ashokdelphia commented:
Really happy to see support for multiple keys and key ids coming along.

It looks like someone could also use this to migrate between algorithms, which is also very welcome.

I'm a little unsure of the behaviour around a missing key id value. I would expect to only need to support that while migrating from an older scheme, to one with key ids.

I think there are three scenarios I would want to support:

- someone is using this with an older configuration: no kid, a single key or secret and single algorithm
- someone moving from an older configuration with no kid, to one with multiple, identified keys while still supporting the tokens issued with no kid; in this case I'd expect to identify in the configuration which key/secret and algorithm to use when the token has no kid
- someone using multiple, identified keys; in this case I'd want to reject tokens that have no kid in the header

I don't think it is desirable to support a fourth case:

- using multiple, identified keys, but allowing an attacker to make tokens that try all of the keys in sequence

nigoroll commented:
Hi @fitodic,

thank you for your comprehensive review, I highly appreciate you taking the time to provide such helpful feedback.

I am going to respond to the overall comments first, then on the detailed comments on code

>     1. The proposed settings schema: If I understood this correctly, each setting accepts a `dict` and a `list`. During the signing process, the first element is used, but there are issued with `dict`s not being ordered in until Python 3.7. Why not split these into multiple settings and drop the use of dicts where it isn't necessary or where order needs to be preserved?

So this only affects `JWT_SECRET_KEY`, the symmetric signing secret setting. For asymmetric crypto, there already is the `JWT_PRIVATE_KEY`/`JWT_PUBLIC_KEY` separation.

As you suggest, one way to solve this would be to add something like a `JWT_SECRET_SIGNING_KEY` which would be required to contain either a scalar or a dict with exactly one member.

As you wrote later on, adding more settings seemed unattractive to me and as this only affected
* users of python < 3.7
* who want to use multiple (named) key ids
* with symmetric signatures

I thought that keeping things simple was the best choice going forward, in particular as the old pythons will eventually become obsolete.

If you think that this is important, I guess our options are:

* to use an `OrderedDict`
* to use another setting like `JWT_SECRET_SIGNING_KEY`

Do you have a preference?

>     2. I suppose the `dict` settings are used to support named keys, but why is this necessary? Is it not enough to handle this via ordering or specifying only the specific keys? I believe adopting a more explicit settings schema would simplify the JWT decode process.

In general, key ids allow to specify which key is to be used, such that at most one verification attempt has to be made. To elaborate:

* In general, trying multiple keys is a waste of cycles, and in particular asymmetric crypto is relatively expensive. Maybe not that much in the realm of django, but, for example my machine does 15492.0 verifies/s for rsa 4096 (`openssl speed rsa`), which makes ~65us per RSA verify in the core openssl code alone. At least in my mind, this is something I would want to avoid

  * this aspect is particularly true verifying JWTs on other systems: Consider a CDN node which has to verify JWTs for every request. For a system which will normally process ~10k req/s on a CPU, one RSA cuts that performance roughly in half, an additional attempt takes it down to about a third.

  * But, true, for this scenario is is more important to produce JWTs with key ids than to verify with optimal efficiency in Django

* naming keys removes guesswork, which I expect to be helpful for analyzing any potential issues. If a key id has been used for signing, that key id needs to verify ok.

> As to the variable names (...)

Yeah, habits of a C developer. BTW, C symbol length does not matter performance wise either, this is more a style habit carried over.

nigoroll commented on docs/index.md:
```diff
@@ -190,19 +214,56 @@ Default is `None`.
+
+For a scalar, this algorithm is used for signing and verification.
+
+For a list, the first element is used for signing and all elements are accepted for verification.
```
I would argue the proposed way is the simplest one, and I would expect it to be sufficient for all practical use cases: I can see no case where you would not want to verify the algorithm used for signing, so I think a list is the right data structure.

nigoroll commented on src/rest_framework_jwt/utils.py:
```diff
@@ -95,8 +95,23 @@ def jwt_get_username_from_payload_handler(payload):
+
+    algo = api_settings.JWT_ALGORITHM
+    if isinstance(algo,list):
+        algo = algo[0]
```
ok

nigoroll commented:
@fitodic I have used separate commits in response to your comments, IMHO these should be squashed before merge

nigoroll commented:
Hi @ashokdelphia ,

> It looks like someone could also use this to migrate between algorithms, which is also very welcome.

Yes, the motivation is to support all kinds of migrations as well as regular key rollover.

If I read your comment correctly, you agree that we should try only a known key by that id if a `kid` header is present.
But you are also saying that we should not always fall back to trying all the keys if there is no `kid` header, right? That would probably require changing the configuration or adding another flag like `INSIST_ON_KID`: Such a flag could be turned on after a transition period to refuse JWTs without a `kid` header.

What do you think?

ashokdelphia commented:
> But you are also saying that we should not always fall back to trying all the keys if there is no `kid` header, right? That would probably require changing the configuration or adding another flag like `JWT_INSIST_ON_KID`: Such a flag could be turned on after a transition period to refuse JWTs without a `kid` header.
> 
> What do you think?

I think you're right that this comes down to how to express it in the config.

I don't think a flag for 'insist on kid' is quite right; if I follow what you intend with that, it sounds like you'd still be looping over all of the keys when someone is mid-transition.

I would be tempted to allow fewer 'shapes' of configuration. Either you have a single key, which doesn't have a kid, or a set of keys each of which must have a kid. In the latter case, you could have an optional config variable for which kid should be used in the case of a token without a kid in the header.

The main case that would disallow would be rotating unidentified keys. I think that's naturally hazardous, and worth excluding so that by the time we're validating a token we always know which algorithm and key / secret should be used.

nigoroll commented:
@ashokdelphia you got a valid point. Having less variants would be advantageous.
Yet IIUC your proposal would make it impossible to transition from multiple, unidentified keys to named keys (with a `kid`). I agree that this setup should be used, but for someone who already finds herself in that situation, there should be a migration option towards the better setup.
IOW, this migration option appears important to me, so being short of any better idea, I will go with `JWT_INSIST_ON_KID` now.

ashokdelphia commented:
> Yet IIUC your proposal would make it impossible to transition from multiple, unidentified keys to named keys (with a `kid`).

Perhaps I'm missing some essential point, but I don't think having multiple unidentified keys is possible at the moment.

I'm not sure it's good to support multiple unidentified keys at all, since I think it would be unsafe in general unless they were either all symmetric or all asymmetric keys. Even then, being able to try a list of keys sounds better for an attacker than for the defence.

If you implement it the way you're thinking will I be able to gracefully transition from a single unidentified key to multiple identified keys without having to allow trying multiple keys, potentially of different algorithms. (In my particular case, I want to move from HS256 to RS512, so I'm naturally concerned about the potential to confuse a public key and a symmetric secret.)

nigoroll commented:
@ashokdelphia You are right, this code does not currently allow for multiple, unidentified keys.
Yet it could be used in an environment where JWTs are generated outside this module, or even outside Django.
In my world, rollover of unidentified keys is common practice, unfortunately, so while I would want to avoid it, I neither want to write code locking in people who find themselves in such environments.
In general, the associated risk is the added validation overhead, as explained [in the second half of this comment](https://github.com/Styria-Digital/django-rest-framework-jwt/pull/33#issuecomment-587354766). This risk will be avoided specifically with `JWT_INSIST_ON_KID` and I do not see any other (security) related risks. In other words, if your policy is to not support multiple, unidentified keys, you will be able to enforce just that.

Let's go through your scenario (please forgive any syntax errors in this mockup code):

* you start off with
```
"JWT_SECRET":"HMAC_KEY"
```

* you add your RSA keypair and specify that you want to use RSA for signing, but still accept HMACs:

```
"JWT_SECRET":"HMAC_KEY",
"JWT_PRIVATE_KEY": {"kid": load_pem_private_key(...)},
"JWT_PUBLIC_KEY": {"kid": load_pem_public_key(...)},
"JWT_INSIST_ON_KID": True,
"JWT_ALGORITHM": ["RS256", "HS256"],
```

* Wait until your HS256 JWTs expire, then remove HS* support:

```
"JWT_SECRET": None,
"JWT_PRIVATE_KEY": {"kid": load_pem_private_key(...)},
"JWT_PUBLIC_KEY": {"kid": load_pem_public_key(...)},
"JWT_INSIST_ON_KID": True,
"JWT_ALGORITHM": "RS256",
```

* Roll over by signing with a new private key, still accepting the old one

```
"JWT_SECRET": None,
"JWT_PRIVATE_KEY": {"nukid": load_pem_private_key(...)},
"JWT_PUBLIC_KEY": {"nukid": load_pem_public_key(...), "kid": load_pem_public_key(...)},
"JWT_INSIST_ON_KID": True,
"JWT_ALGORITHM": "RS256",
```

ashokdelphia commented:
@nigoroll: Thanks for the detailed example. I think I was misunderstanding how INSIST_ON_KID would work. In step 2 there, it sounds like INSIST_ON_KID would be set, but we'd still be accepting symmetric tokens with no kid, which would indeed allow what I was asking about.

I'll take a closer look at the implementation later; thanks.

nigoroll commented:
@ashokdelphia I still have not written the `INSIST_ON_KID` code so there is nothing for you to look at more closely yet.
I should be ready in an hour or so, please check back then.

nigoroll commented:
I have force-pushed an update which addresses feedback:
* To solve the dict ordering issue, we really do not need to do much except to advise users of python versions < 3.7 to use an `OrderedDict` instead of an ordinary dict
* `INSIST_ON_KID` has been added, see previous notes for detailed discussion.

ashokdelphia commented on changelog.d/33.feature.md:
```diff
@@ -0,0 +1,34 @@
+
+  Existing code made key rollovers or algorithm changes hard and
+  basically required a breaking change: Once any of `JWT_ALGORITHM`,
+  `JWT_SECERT_KEY`, or `JWT_PRIVATE_KEY`/`JWT_PUBLIC_KEY` were
```
There's a type in `JWT_SECERT_KEY`, and in two spots just below.

nigoroll replied:
well spotted, thank you!

ashokdelphia commented:
> I have force-pushed an update which addresses feedback:

Thanks for being defensive about the potential for crossing symmetric and asymmetric algorithms.

I also appreciate how you've kept that separate, and also allowed `INSIST_ON_KID` to permit `kid`-less tokens as long as key ids are not defined in the relevant config. That will make this a good fit for a migration I need to do shortly.

nigoroll commented:
force-pushed adressing @ashokdelphia 's feedback

fitodic commented on src/rest_framework_jwt/utils.py:
```diff
@@ -107,17 +121,54 @@ def jwt_decode_token(token):
+        if kid:
+            try:
+                keys = keys[kid]
+            except:
```
I assume this can only raise a `KeyError`, but the current implementation catches everything, including `KeyboardInterrupt`s. I recommend that the exceptions are explicitly stated.

nigoroll replied:
right

nigoroll commented on docs/index.md:
```diff
@@ -189,6 +189,38 @@ This package uses the JSON Web Token Python implementation, [PyJWT](https://gith
+
+* when a list, all of the elements will be accepted for verification of JWTs.
+
+When a scalar, this secret is used for signing and verification.
```
good point, docs polished

nigoroll commented on docs/index.md:
```diff
@@ -189,6 +189,38 @@ This package uses the JSON Web Token Python implementation, [PyJWT](https://gith
+
+When a scalar, this secret is used for signing and verification.
+
+When a list or dict, the first element/value is used for signing.
```
good point, docs polished

nigoroll commented on docs/index.md:
````diff
@@ -200,22 +232,65 @@ Default is `None`.
+{ "kid": key }
+```
+
+The scalar or the dict value must be in any [private key format supported by PyJWT](https://github.com/jpadilla/pyjwt/blob/master/docs/algorithms.rst), for example of the types
````
good point, thank you

nigoroll commented on docs/index.md:
````diff
@@ -200,22 +232,65 @@ Default is `None`.
+  { "kid1": key1, "kid2": key2, ... }
+  ```
+
+  If a JWT is to be verified which contains a key id (kid header), only
````
yes, my wording was bad. Thank you

nigoroll commented on docs/index.md:
```diff
@@ -200,22 +232,65 @@ Default is `None`.
+
+For a scalar, this algorithm is used for signing and verification.
+
+For a list, the first element is used for signing and all elements are accepted for verification.
```
makes sense

nigoroll commented on tests/views/test_authentication.py:
```diff
@@ -162,3 +196,138 @@ def test_auth_cookie_settings(
+    response = call_auth_endpoint("username", "password")
+    token = response.json()["token"]
+
+    monkeypatch.setattr(api_settings, "JWT_SECRET_KEY", ["key2", "key1"])
```
Yes, that is the intention.

nigoroll commented on tests/views/test_authentication.py:
```diff
@@ -162,3 +196,138 @@ def test_auth_cookie_settings(
+    monkeypatch, user, call_auth_endpoint, rsa_keys
+):
+
+    if not rsa_keys:
+        skip("no rsa available")
+        return
```
I would leave the decision with you, but my personal preference is to keep the number of additional dependencies as small as possible.

fitodic replied:
That's my personal preference as well, but I've been down the rabbit hole called Python packaging enough times to care about the UX :laughing: 

That said, I'd prefer our users have a streamlined experience when it comes to installing and using this library. I wouldn't want to subject them to searching through requirements, the documentation or GitHub issues only to find out that there's a caveat to using a certain feature. It's possible that we'll have to warn them to install `openssl` before installing or upgrading, but if the feature you're introducing is managed via settings, let's leave it at that.

The other upside to this approach is that `PyJWT` implicitly defines which versions of `cryptography` it's compatible with. That means that neither our users nor we have to worry about it unless we have conflicting versions somewhere, but that's something only the users can fix.

nigoroll replied:
ok, see 73536b899827f9509e0cc5021f9f84b077c0dc38

nigoroll commented on tests/views/test_authentication.py:
```diff
@@ -162,3 +196,138 @@ def test_auth_cookie_settings(
+
+
+def test_multi_algo(
+        monkeypatch, user, call_auth_endpoint, rsa_keys
```
yeah

nigoroll commented on tests/views/test_authentication.py:
```diff
@@ -162,3 +196,138 @@ def test_auth_cookie_settings(
+        api_settings, "JWT_PUBLIC_KEY", rsa_keys["public"]
+    )
+
+    sk = OrderedDict(hash1="one", hash2="two")
```
We come from different traditions and I appreciate your patience with me, I will try to adjust to the project's style.

nigoroll commented:
@fitodic Apologies for the long delay on my side and thank you very much for your detailed feedback.
I hope to have addressed all of it. At this point, here is a travis build error which seems unrelated, but I will try to fix or work around it to get ahead.
Please note that I have kept some intermediate commits with the intent of making your review easier, but I would suggest we squash some before a possible merge.
Again, thank you!

nigoroll commented:
so with `pytest`, all tests succeed, but I can reproduce the pytest failures via `tox` locally as seen with travis. If anyone has any input?

nigoroll commented:
wrt https://github.com/Styria-Digital/django-rest-framework-jwt/pull/33#issuecomment-616475204 I am totally lost. Will tend to something else any would appreciate help from a python wizard

fitodic commented:
> @fitodic Apologies for the long delay on my side and thank you very much for your detailed feedback.
> I hope to have addressed all of it. At this point, here is a travis build error which seems unrelated, but I will try to fix or work around it to get ahead.
> Please note that I have kept some intermediate commits with the intent of making your review easier, but I would suggest we squash some before a possible merge.
> Again, thank you!

There's really no need to apologize. These are stressful times for all of us and I really appreciate your continued involvement. These things take time, especially in these circumstances, so there's really no rush. Take care of yourself and your loved ones and we'll bring this to the finish line in no time. :slightly_smiling_face: 
Don't worry about the commits or waste time merging them. If that is what you wish, I'll squash them during the merge process.

> so with `pytest`, all tests succeed, but I can reproduce the pytest failures via `tox` locally as seen with travis. If anyone has any input?
> This happens equally on the master branch, so whatever changed seems to be unrelated to this PR.
> 
> (edit: clarify that the tox/pytest issue is unrelated to this PR)

I'm a bit lost. You _can_ or _can't_ reproduce this issue? As you mentiond, this probably isn't related to this PR as it's [failing in the master branch](https://travis-ci.org/github/Styria-Digital/django-rest-framework-jwt/jobs/668676853) as well. I'll have a look and get back to you as soon as I have something concrete.

nigoroll commented:
@fitodic thank you and I also wish you and really everybody to stay safe.
Regarding the squashes: Yes, I would like to tidy up the commits when we are ready or you can do it also. If you just want to merge when you feel we are ready, please do whatever you find appropriate or let me know to do a final squash.
Regarding the jenkins/tox failure:
* running `pytest` locally succeeds
* running `tox` locally fails for a failing `pytest`

This is _is_ reproducible locally and not just happening with Jenkins.

nigoroll commented:
After another minor change it seems we are getting a green light from travis now. I will squash commits.

nigoroll commented:
squash done

fitodic commented on docs/index.md:
````diff
@@ -189,6 +189,39 @@ This package uses the JSON Web Token Python implementation, [PyJWT](https://gith
+
+* When a scalar, this secret is used for signing and verification.
+
+*NOTE: For python < 3.7, use a `collections.OrderedDict` object*, e.g.:
+
+  ```python
+  from collections import OrderedDict
+
+  JWT_AUTH["JWT_SECRET_KEY"] = OrderedDict(kid1=key1, kid2=key2, ...)
+  ```
+
+(The first) `JWT_SECRET_KEY` is only used for signing if (the first)
+`JWT_ALGORITHM` is `HS*`, otherwise `JWT_PRIVATE_KEY` is used.
+
+`JWT_SECRET_KEY`(s) is/are only used for verification of JWTs with
+`alg` matching `HS*`
````
What do you say we move this block up under the `When a dict` point, just before the second `When a list` point? That way we'll reduce the risk of users skipping it by accident.

nigoroll replied:
makes sense

nigoroll commented on changelog.d/33.feature.md:
```diff
@@ -0,0 +1,42 @@
+  Setting `JWT_INSIST_ON_KID: True` avoids this fallback and requires
+  any JWT to be validated to carry a key id _if_ key IDs are used
+
+  *NOTE: For python < 3.7, use a `collections.OrderedDict` object
+  instead of a dict
```
sorry and thank you for spotting it.

nigoroll commented:
@fitodic Sorry for having missed some conversations, gh hid them from me by default and I assumed that it would only do that for resolved conversations. Looking at them now

nigoroll commented on docs/index.md:
```diff
@@ -200,22 +232,65 @@ Default is `None`.
+When key IDs are used (`JWT_SECRET_KEY` and/or `JWT_PUBLIC_KEY` given
+as a dict assigning key IDs to keys), insist that JWTs to be validated
+have a `kid` header with a defined key.
+
```
right, thx

fitodic commented:
> @fitodic Sorry for having missed some conversations, gh hid them from me by default and I assumed that it would only do that for resolved conversations. Looking at them now

Don't worry about it, happens to all of us. That's why I started reviewing everything anew from the `/files` endpoint some time ago. At least GitHub has the `Mark as viewed` checkbox next to each file :laughing:

nigoroll commented on tests/views/test_authentication.py:
```diff
@@ -2,16 +2,50 @@
+    secret = {}
+    public = {}
+
+    for i in range(1, nkeys):
```
that was a stupid mistake on my side, this was meant to generate two keys.

nigoroll commented:
@fitodic I feel bad about wasting your time with these little details, thank you for your thorough work.
I hope all comments are addressed now.

fitodic commented:
> @fitodic I feel bad about wasting your time with these little details, thank you for your thorough work.
> I hope all comments are addressed now.

After all your hard work, it's the least I can do. :slightly_smiling_face:

fitodic commented:
@nigoroll I've squashed the commits as you've requested. This feature definitely improves the overall quality of this library so once again, thank you for your contribution and patience. :slightly_smiling_face:

nigoroll commented:
@fitodic it has been a pleasure working with you

fitodic commented:
> @fitodic it has been a pleasure working with you

It has been a pleasure working with you too :slightly_smiling_face:  The new release is available on [PyPI](https://pypi.org/project/drf-jwt/1.16.0/), and the [documentation](https://styria-digital.github.io/django-rest-framework-jwt/) has been updated.

nigoroll commented:
thx. Seeing there html-render of the docs I noticed that we could look after some indentation fixes still and maybe give the settings own paragraphs. I might get back to that
